azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,325 @@
+                if "operationId" in publish_action.publish_topic:
+                    self._polling_timer.cancel()
+                    self._polling_timer = None
+            self._trig_wait(registration_result)
+        elif registration_result.status == "assigned":
+            self._trig_complete(registration_result)
+        else:
+            self._trig_error(registration_result)
+
+    def clear_and_disconnect(self, event_data):
+        self._query_timer = None
+        self._polling_timer = None
+        unsubscribe_action = UnsubscribeAction(
+            constant.SUBSCRIBE_TOPIC_PROVISIONING, callback=self.on_unsubscribe_completed
+        )
+        self._request_response_provider.disconnect(
+            callback=self.on_disconnect_completed, pre_disconnect_actions=[unsubscribe_action]
+        )
+        # TODO : Do I need to wait for disconnect and then call the cancel callback?
+        callback = self._cancel_callback
+
+        if callback:
+            self._cancel_callback = None
+            callback()
+
+    def _call_error(self, event_data):
+        logger.info("Failed register from polling machine")
+        error = event_data.args[0]
+        callback = self._register_callback
+
+        if callback:
+            self._register_callback = None
+            callback(error)
+
+    def _call_complete(self, event_data):
+        logger.info("Complete register from polling machine")
+        registration_result = event_data.args[0]
+        callback = self._register_callback
+
+        if callback:
+            self._register_callback = None
+            callback(registration_result)
+        # TODO : Should disconnect be by default called here ?
+
+    def _set_query_timer(self):
+        def time_up_query():
+            logger.error("Time is up for query timer")
+            # TODO TimeoutError causing problems
+            self._trig_error(ValueError("Time is up for query timer"))
+
+        self._query_timer = Timer(constant.DEFAULT_TIMEOUT_INTERVAL / 1000, time_up_query)
+        self._query_timer.start()
+
+    def _wait_for_interval(self, event_data):
+        def time_up_polling():
+            logger.info(
+                "Done waiting for polling interval of %s ms", str(constant.DEFAULT_POLLING_INTERVAL)
+            )
+
+        self._polling_timer = Timer(constant.DEFAULT_POLLING_INTERVAL / 1000, time_up_polling)
+        logger.info("Waiting for " + str(constant.DEFAULT_POLLING_INTERVAL) + " ms")
+        self._polling_timer.run()  # This is waiting for that polling interval
+        self._trig_poll(event_data)
+
+    def on_connect_completed(self):
+        logger.info("on_connect_completed for DPS")
+        if self.on_connected:
+            self.on_connected("connected")
+
+    def on_disconnect_completed(self):
+        logger.info("on_disconnect_completed for DPS")
+        if self.on_disconnected:
+            self.on_disconnected("disconnected")
```

Comment:
Calling a callback should have a try/except around it, since it's calling code unknown to this layer.

Reply:
+1.  I've spent lots of time in the Node SDK tracking down callbacks (actually handlers) that had exceptions.  Without a try/except here, stuff breaks and there's no explanation to help you debug.  

---
In reply to: [277787301](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277787301) [](ancestors = 277787301)